src/org/rascalmpl/library/analysis/m3/Core.rsc

```diff
@@ -87,14 +88,42 @@ private value diff(value v1, value v2) { throw "can\'t differentiate non-collect
 	Generic function to compose the annotations of a set of M3s.
 }
 @memo
-M3 composeM3(loc id, set[M3] models) = modifyM3(id, toList(models), compose);
+M3 composeM3(loc id, set[M3] models) {
+	M3 comp = m3(id);
+
+	comp.declarations = {*model.declarations | model <- models};
+	comp.types = {*model.types | model <- models};
+	comp.uses = {*model.uses | model <- models};
+	comp.containment = {*model.containment | model <- models};
+	comp.messages = [*model.messages | model <- models];
+	comp.names = {*model.names | model <- models};
+	comp.documentation = {*model.documentation | model <- models};
+	comp.modifiers = {*model.modifiers | model <- models};
```

Comment:
nice, but sad that this is needed.

I think you could also write in in a generic way, where you would only have to iterate over the models once, and use the `Node::getKeywordParameters` and `Node::setKeywordParameters`? should still be about as fast I would expect.

Reply:
The generic way would not have my preference here, this code is easier tractable, for correctness and declarativeness and type safety.

Reply:
okay, that's right, I didn't like the duplicity of this code, just to splice the stuff.. I think we really need a way just to splice (flatsplice?) the lists. so something like `comp.messages = *models.messages`

Reply:
What about Set::union?

Reply:
that doesn't work, since it's a selection of a field in a set. you would still have to write the comprehension to do the map.

Reply:
Same reasoning on my side: the code could be (and was) generic, but since the M3 ADT is not expected to change every other day, I feel like the explicit version is clearer. Also, having a generic compose/diff would make people expect that any kind of M3 model could be composed in this way; which might not always be true.